MCP23SXX I/O Expander - SPI

SenexCrenshaw opened the pull request:
## Description:
SPI versions of the MCP230xx I/O Expander. Was thinking I could write a base class and then modify the devices to match but i cant test all the devices and i figured to leave his code alone. I did however mirror it as much as possible for folks to not relearn different approaches if they need to dig into the code.

I also added 10MHZ to the SPI speeds. theses devices can easily handle that even at their lowest min voltage.

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#613

## Checklist:
  - [X] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

SenexCrenshaw committed:
MCP23SXX I/O Expander - SPI

SenexCrenshaw committed:
MCP23SXX I/O Expander - SPI

SenexCrenshaw committed:
MCP23SXX I/O Expander - SPI

SenexCrenshaw committed:
MCP23SXX I/O Expander - SPI

SenexCrenshaw committed:
Travis Fixes

SenexCrenshaw committed:
Travis Fixes

SenexCrenshaw committed:
Travis Fixes

SenexCrenshaw committed:
Travis Fixes

SenexCrenshaw committed:
Build Fix

SenexCrenshaw committed:
Build Fix

SenexCrenshaw committed:
Build Fix

SenexCrenshaw committed:
Build Fix

SenexCrenshaw committed:
Travis fix

SenexCrenshaw committed:
Travis fix

SenexCrenshaw committed:
Travis fix

SenexCrenshaw committed:
Travis fix

Sabesto commented:
Confirmed working on a PRODINo ESP32 Ethernet with mcp23s08

AdamN1 commented:
@Sabesto will you resolve conflicts in this PR so we can marge this to master?

SenexCrenshaw commented:
Sorry. for the delay! I will work on this today/tomorrow.

Sabesto commented:
> Sorry. for the delay! I will work on this today/tomorrow.
> 
> I have added a mcp3008 module in esphome as well and added a reference voltage line. Basically it returns the voltage value mimicking the ADC component in esphome ( value_v = (value_v * this->reference_voltage_) / 1023;) Ill add it to this one as well.

One PR for each component would make more sense

> Last thing. What exactly do you want to test? A yaml file and that the code generated compiles? If so where do you want this? any examples?

Look at my PR: esphome/esphome#1324 specifically tests/test1.yaml

Sabesto commented on esphome/components/mcp23S08/__init__.py:
```diff
@@ -0,0 +1,54 @@
+
+CONF_DEVICEADDRESS = "deviceaddress"
+
+mcp23S08_ns = cg.esphome_ns.namespace('mcp23S08')
```
I think namespace should be lowercase, mcp23s08, not mcp23S08

SenexCrenshaw replied:
ill have a more robust patch in today or tomorrow. clearly i wont be doing the ref voltage thing since these are digital and not ADC....

SenexCrenshaw committed:
Code cleanup
Removed capital S where found
changed pin config option to mcp23sxx_id:
updated test1.yaml

SenexCrenshaw committed:
Code cleanup
Removed capital S where found
changed pin config option to mcp23sxx_id:
updated test1.yaml

SenexCrenshaw committed:
Lint fixes

SenexCrenshaw committed:
Lint fixes

SenexCrenshaw committed:
Update init

SenexCrenshaw committed:
Added CODEOWNERS

SenexCrenshaw committed:
Added  DATA_RATE_10MHZ = 10000000 to spi

jesserockz commented on esphome/components/mcp23s08/mcp23s08.cpp:
```diff
@@ -0,0 +1,121 @@
+
+void MCP23S08::dump_config() {
+  ESP_LOGCONFIG(TAG, "MCP23S08:");
+  ESP_LOGCONFIG(TAG, "  CS Pin: %u", this->cs_->get_pin());
```
```suggestion
  LOG_PIN("  CS Pin: ", this->cs_);
```

SenexCrenshaw replied:
Ok

jesserockz commented on esphome/components/mcp23s17/__init__.py:
```diff
@@ -0,0 +1,59 @@
+    yield spi.register_spi_device(var, config)
+
+
+CONF_MCP23S17 = 'mcp23s17'
+CONF_MCP23S17_ID = 'mcp23s17_id'
+
+mcp23S17_OUTPUT_PIN_SCHEMA = cv.Schema({
+    cv.GenerateID(CONF_MCP23S17_ID): cv.use_id(mcp23S17),
+    cv.Required(CONF_NUMBER): cv.int_,
+    cv.Optional(CONF_MODE, default="OUTPUT"): cv.enum(mcp23S17_GPIO_MODES, upper=True),
+    cv.Optional(CONF_INVERTED, default=False): cv.boolean,
+})
+mcp23S17_INPUT_PIN_SCHEMA = cv.Schema({
+    cv.Required(CONF_MCP23S17_ID): cv.use_id(mcp23S17),
+    cv.Required(CONF_NUMBER): cv.int_range(0, 15),
+    cv.Optional(CONF_MODE, default="INPUT"): cv.enum(mcp23S17_GPIO_MODES, upper=True),
+    cv.Optional(CONF_INVERTED, default=False): cv.boolean,
+})
+
+
+@pins.PIN_SCHEMA_REGISTRY.register(CONF_MCP23S17_ID,
+                                   (mcp23S17_OUTPUT_PIN_SCHEMA, mcp23S17_INPUT_PIN_SCHEMA))
+def mcp23S17_pin_to_code(config):
+    parent = yield cg.get_variable(config[CONF_MCP23S17_ID])
+    yield mcp23S17GPIOPin.new(parent, config[CONF_NUMBER], config[CONF_MODE], config[CONF_INVERTED])
```
The `..._PIN_SCHEMA`s do not use an `_ID` suffix

```suggestion
CONF_MCP23S17 = 'mcp23s17'

mcp23S17_OUTPUT_PIN_SCHEMA = cv.Schema({
    cv.GenerateID(CONF_MCP23S17): cv.use_id(mcp23S17),
    cv.Required(CONF_NUMBER): cv.int_,
    cv.Optional(CONF_MODE, default="OUTPUT"): cv.enum(mcp23S17_GPIO_MODES, upper=True),
    cv.Optional(CONF_INVERTED, default=False): cv.boolean,
})
mcp23S17_INPUT_PIN_SCHEMA = cv.Schema({
    cv.Required(CONF_MCP23S17): cv.use_id(mcp23S17),
    cv.Required(CONF_NUMBER): cv.int_range(0, 15),
    cv.Optional(CONF_MODE, default="INPUT"): cv.enum(mcp23S17_GPIO_MODES, upper=True),
    cv.Optional(CONF_INVERTED, default=False): cv.boolean,
})


@pins.PIN_SCHEMA_REGISTRY.register(CONF_MCP23S17,
                                   (mcp23S17_OUTPUT_PIN_SCHEMA, mcp23S17_INPUT_PIN_SCHEMA))
def mcp23S17_pin_to_code(config):
    parent = yield cg.get_variable(config[CONF_MCP23S17])
    yield mcp23S17GPIOPin.new(parent, config[CONF_NUMBER], config[CONF_MODE], config[CONF_INVERTED])
```

jesserockz commented on esphome/components/mcp23s08/__init__.py:
```diff
@@ -0,0 +1,59 @@
+    yield spi.register_spi_device(var, config)
+
+
+CONF_MCP23S08 = 'mcp23s08'
+CONF_MCP23S08_ID = 'mcp23s08_id'
+
+mcp23S08_OUTPUT_PIN_SCHEMA = cv.Schema({
+    cv.GenerateID(CONF_MCP23S08_ID): cv.use_id(mcp23S08),
+    cv.Required(CONF_NUMBER): cv.int_,
+    cv.Optional(CONF_MODE, default="OUTPUT"): cv.enum(mcp23S08_GPIO_MODES, upper=True),
+    cv.Optional(CONF_INVERTED, default=False): cv.boolean,
+})
+mcp23S08_INPUT_PIN_SCHEMA = cv.Schema({
+    cv.GenerateID(CONF_MCP23S08_ID): cv.use_id(mcp23S08),
+    cv.Required(CONF_NUMBER): cv.int_range(0, 7),
+    cv.Optional(CONF_MODE, default="INPUT"): cv.enum(mcp23S08_GPIO_MODES, upper=True),
+    cv.Optional(CONF_INVERTED, default=False): cv.boolean,
+})
+
+
+@pins.PIN_SCHEMA_REGISTRY.register(CONF_MCP23S08_ID,
+                                   (mcp23S08_OUTPUT_PIN_SCHEMA, mcp23S08_INPUT_PIN_SCHEMA))
+def mcp23S08_pin_to_code(config):
+    parent = yield cg.get_variable(config[CONF_MCP23S08_ID])
+    yield mcp23S08GPIOPin.new(parent, config[CONF_NUMBER], config[CONF_MODE], config[CONF_INVERTED])
```
The `..._PIN_SCHEMA`s do not use an `_ID` suffix

```suggestion
CONF_MCP23S08 = 'mcp23s08'

mcp23S08_OUTPUT_PIN_SCHEMA = cv.Schema({
    cv.GenerateID(CONF_MCP23S08): cv.use_id(mcp23S08),
    cv.Required(CONF_NUMBER): cv.int_,
    cv.Optional(CONF_MODE, default="OUTPUT"): cv.enum(mcp23S08_GPIO_MODES, upper=True),
    cv.Optional(CONF_INVERTED, default=False): cv.boolean,
})
mcp23S08_INPUT_PIN_SCHEMA = cv.Schema({
    cv.GenerateID(CONF_MCP23S08): cv.use_id(mcp23S08),
    cv.Required(CONF_NUMBER): cv.int_range(0, 7),
    cv.Optional(CONF_MODE, default="INPUT"): cv.enum(mcp23S08_GPIO_MODES, upper=True),
    cv.Optional(CONF_INVERTED, default=False): cv.boolean,
})


@pins.PIN_SCHEMA_REGISTRY.register(CONF_MCP23S08,
                                   (mcp23S08_OUTPUT_PIN_SCHEMA, mcp23S08_INPUT_PIN_SCHEMA))
def mcp23S08_pin_to_code(config):
    parent = yield cg.get_variable(config[CONF_MCP23S08])
    yield mcp23S08GPIOPin.new(parent, config[CONF_NUMBER], config[CONF_MODE], config[CONF_INVERTED])
```

SenexCrenshaw replied:
Ok!

jesserockz commented on esphome/components/mcp23s17/mcp23s17.cpp:
```diff
@@ -0,0 +1,126 @@
+
+void MCP23S17::dump_config() {
+  ESP_LOGCONFIG(TAG, "MCP23S17:");
+  ESP_LOGCONFIG(TAG, "  CS Pin: %u", this->cs_->get_pin());
```
```suggestion
  LOG_PIN("  CS Pin: ", this->cs_);
```

jesserockz commented on tests/test1.yaml:
```diff
@@ -799,6 +809,24 @@ esp32_touch:
+  - platform: gpio
+    name: "MCP23S08 Pin #1"
+    pin:
+      mcp23s08_id: mcp23s08_hub
```
```suggestion
      mcp23s08: mcp23s08_hub
```

jesserockz commented on tests/test1.yaml:
```diff
@@ -799,6 +809,24 @@ esp32_touch:
+  - platform: gpio
+    name: "MCP23S17 Pin #1"
+    pin:
+      mcp23s17_id: mcp23s17_hub
```
```suggestion
      mcp23s17: mcp23s17_hub
```

jesserockz commented on tests/test1.yaml:
```diff
@@ -1323,6 +1351,23 @@ climate:
 switch:
+  - platform: gpio
+    name: "MCP23S08 Pin #0"
+    pin:
+      mcp23s08_id: mcp23s08_hub
+      # Use pin number 0
+      number: 0
+      # One of INPUT, INPUT_PULLUP or OUTPUT
+      mode: INPUT_PULLUP
+      inverted: False
```
```suggestion
    pin:
      mcp23s08: mcp23s08_hub
      # Use pin number 0
      number: 0
      mode: OUTPUT
      inverted: False
```

jesserockz commented on tests/test1.yaml:
```diff
@@ -1323,6 +1351,23 @@ climate:
+  - platform: gpio
+    name: "MCP23S17 Pin #0"
+    pin:
+      mcp23s17_id: mcp23s17_hub
```
```suggestion
      mcp23s17: mcp23s17_hub
```

jesserockz reviewed (changes requested):
A few changes to line up with other pin devices please

SenexCrenshaw committed:
Update esphome/components/mcp23s08/mcp23s08.cpp

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

SenexCrenshaw committed:
Update esphome/components/mcp23s08/__init__.py

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

SenexCrenshaw committed:
Update tests/test1.yaml

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

SenexCrenshaw committed:
Update tests/test1.yaml

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

SenexCrenshaw committed:
Update tests/test1.yaml

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

SenexCrenshaw committed:
Update tests/test1.yaml

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

SenexCrenshaw committed:
Update esphome/components/mcp23s17/__init__.py

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

SenexCrenshaw committed:
Update esphome/components/mcp23s17/mcp23s17.cpp

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

SenexCrenshaw commented:
@jesserockz Thanks for the tips. I committed your changes and appreciate the help!

jesserockz reviewed (approved):
Looks good, thanks

Melkor82 commented:
@SenexCrenshaw can you make the spi speed an option to configure? Now it is configured as DATA_RATE_10MHZ but i would like to try to reach bigger distances using lower speed rate

SenexCrenshaw commented:
@Melkor82 I dont believe that can be changed from a configuration. Id suggest downloading the git repo and modifying to your needs. 
@jesserockz  @glmnet  - Am I correct? SPI speeds cant be changed, right?

glmnet commented:
I don't know of any spi component which would allow to do that. Sounds like a core requirement to support this

makkie2002 commented:
> Confirmed working on a PRODINo ESP32 Ethernet with mcp23s08

That is really fantastisch! I was hoping for this integration. I also have a Prodino. Can you please post an example of a working configuration file so that I can get started. Thanks!

SenexCrenshaw commented:
This was added to the latest release. Here are the docs
https://esphome.io/components/mcp23sxx
